src/librustc/ty/print/mod.rs

```diff
@@ -0,0 +1,318 @@
+use hir::map::{DefPathData, DisambiguatedDefPathData};
```

Comment:
> * I don't _really_ understand the role of these traits

The PR description *tries* to explain it but it probably fails to do so fully.

I need a better name than `Printer`/`print_*` for the general case, it's *effectively* a serialization scheme where the primitives are paths and several typesystem entities, only `PrettyPrinter` is *really* about "printing".

There might not even any point to having `Printer` (we can take the little logic it has and make it available some other way), and we can move its interface into `PrettyPrinter`, which can actually have defaults for (almost) everything (i.e. you can get `FmtPrinter`, or something similar, very easily, and then you can just customize the parts you care about).

> * When is it appropriate to create a new printer (I've seen a few in this PR, but I can't quite tell what for)

It's not clear to me right now too, other than to replace the old `ItemPathBuffer` trait.

In #57967 you can see I created a `Printer` in `rustc_codegen_utils::symbol_names::mw` which relies on the associated types, but almost everything else is `PrettyPrinter` and that you would create whenever `FmtPrinter` is not enough.

I would even take the `RegionHighlightMode` out of `FmtPrinter` and move it to wherever most of the lifetime placeholder code lives, but I felt that was even more churn to add to this PR.

> * When would it make sense to (e.g.) add a new `print_my_type` method to `Printer`?

When you want to overload it in an implementation, without having to reimplement printing for whatever happens to contain it (e.g. paths, types, etc.). This is usually the case for all "typesystem entities".

For example, @varkor will have to add a `print_const` method that takes a `ty::LazyConst<'tcx>` (unless @oli-obk wants to rename it again), because printers will want to choose how they print constants without having to handle that in a few separate places.

Reply:
Writing up these notes in a comment would be great